### ISSUE TITLE
Allow specifying a `to` address in the flash mint

### ISSUE THREAD

**Issue by moodysalem** [WETH10/WETH10#56]
Instead of calling the msg sender, mint to the given `to` address and call it

**Comment by albertocuestacanada** [comment]
We did that for fyDai, only to discover that you could flash mint to someone that traded based on balances (such as an AMM) and exploit it.

**Comment by albertocuestacanada** [comment]
Given the general reluctance to flash mint in general I'm not too keen in adding more complexity to the function.

Could this be exploited to flash mint on an unsuspecting contract, say an AMM, that trades based on balances?

I'm thinking it couldn't, because the receiver would have to implement the `executeOnFlashMint` function for the flash mint to be successful.

**Comment by moodysalem** [comment]
> Given the general reluctance to flash mint in general I'm not too keen in adding more complexity to the function.
> 
> Could this be exploited to flash mint on an unsuspecting contract, say an AMM, that trades based on balances?
> 
> I'm thinking it couldn't, because the receiver would have to implement the `executeOnFlashMint` function for the flash mint to be successful.

I made a separate issue about removing that interface because flash minting to an AMM is not a concern but rather a useful feature. eg I want to do some arbitrage with WETH I dont have.

**Comment by albertocuestacanada** [comment]
Take into account that a flash mint is:
 1. `mint`
 2. `execute`
 3. `burn`

By forcing the flash mint to go to msg.sender, it means that msg.sender must go through the usual functions to interact with a third contract (i.e. some function that does `transferFrom` and puts the control in the hands of the third contract).

If you allow to execute the flash mint on any arbitrary contract, you are also allowing WETH10.sol to burn WETH from that arbitrary contract at the end of the flash mint, without consent from that arbitrary contract. WETH10 is opt-in, so that can be coded safely, but that is unusual behaviour for an ERC20 token.

Any factory contract that accepts ERC20 tokens must either deal with all of them as if they could flash mint and burn directly, code a special case for WETH10, or reject it.

UniswapV2 might be able to deal with this, or maybe not.
1. Start with `d` Dai.
2. Mint `w` WETH into a WETH/Dai pair
3. Do a small `swap` to `_update` the reserves, incorporating the small trade but also the `w` WETH
4. WETH is now cheaper than it was, so with the `d` Dai buy as much WETH from the Pair as you can.
5. WETH10.sol will `burn` `w` WETH directly from the Pair.
6. You bought WETH from the Pair at an artificially depressed price, with the Pair bearing the losses.

I'm not sure if UniswapV2 could be exploited this way, but others could, which would lead to WETH10 not being adopted.

**Comment by Austin-Williams** [comment]
I agree with only flash minting to `msg.sender`. It reduces the attack surface quite a bit.

If we _do_ allow flash minting to third party accounts, then we should also have the WETH10 contract pass the value of `msg.sender` to the target contract [on line 120](https://github.com/WETH10/WETH10/blob/main/contracts/WETH10.sol#L120). That way, the target contract would know what address initiated the flash mint and could decide whether or not to proceed.

**Comment by moodysalem** [comment]
Rather than burning from the `to` account, simply check the invariant after flash minting and calling
i.e.
`address(this).balance >= totalSupply()`

the recipient of the flash mint just has to ensure this variant holds true. however there is no way to reduce total supply / return the minted WETH, without a `burn` method

you can add a `burn` method that simply reduces the caller's WETH balance without withdrawing any ETH for this

tbh i'm not entirely sure about this in terms of security, but i think it's logically correct if you wanted to be able to flashMint to a given address

**Comment by Austin-Williams** [comment]
> Rather than burning from the to account, simply check the invariant after flash minting/calling the to address
i.e.
`address(this).balance >= totalSupply()`

So I've thought about this before (it was actually in the original implementation of flash-mintable tokens), but there are two reasons I'm not a fan of this approach.

First, there is/was a separate issue related to `totalSupply()` not returning the total number of WETH in existence. So if this approach is taken, it is 100% necessary to make sure that the `totalSupply` is tracked properly.

Second, and more importantly, requiring that `address(this).balance >= totalSupply()` is true at the end of the flash mint means that users cannot do multiple nested flash mints.

E.g., suppose the Compound platform started using a WETH10 flash mint as part of one of its internal functions. And then suppose a user does a flash mint and then -- during their flash mint -- interacts with Compound. In that situation, Compound's internal flash mint will revert when doing the `address(this).balance >= totalSupply()` (even if Compound does everything right). The `address(this).balance >= totalSupply()` check (at the end of Compound's flash mint) will fail because the original user's flash mint hasn't been paid back yet.

Tl;Dr: The `address(this).balance >= totalSupply()` check would break flash mint composability.

**Comment by albertocuestacanada** [comment]
Ooooh, now you have given me a nice challenge.

I'll revert `totalSupply` to be just `address(this).balance` and merge #63, which by then would contain the uncontroversial feedback.

Then in a separate PR I'll do flash minting to `to`, `burn` and `burnFrom`,  and an array of `supplyTargets` to meet at the end of each flag mint so that you can nest in and `pop` them in the way out.

We might or not merge that one, but it will be fun.

**Comment by albertocuestacanada** [comment]
@moodysalem, @Austin-Williams - There is an implementation of this pattern in #65.

@dmihal came with the flashMint + flashBurn pattern a while ago, but without the idea of minting to an arbitrary address it wasn't needed and we discarded it.

**Comment by Austin-Williams** [comment]
> I don't see minting into arbitrary addresses as dangerous. It's the same as a mint to self and a transfer.

It is different because it allows _anyone_ to cause the WETH10 contract to call anyone else's `executeOnFlashMint` function. It requires that all implementors of an `executeOnFlashMint` function add access controls to that function.

To allow for these access controls (assuming you do end up taking this approach), the `executeOnFlashMint` function (https://github.com/WETH10/WETH10/pull/65/files#diff-0f1da5038a4b3a4a0936420a907f6f8e5826984e4c3647fa74938438111116deR101) should accept a `_caller` parameter, and  WETH10 should pass `msg.sender` as the `_caller` [here](https://github.com/WETH10/WETH10/pull/65/files#diff-0f1da5038a4b3a4a0936420a907f6f8e5826984e4c3647fa74938438111116deR101).

(This is similar to how Uniswapv2 flash swaps work)

**Comment by Austin-Williams** [comment]
All of that being said, I would recommend keeping it simple and not allowing the flash minter to specify a `to` address. It keeps the code simpler, and makes it easier and safer for people implementing an `executeOnFlashMint` function on their contracts.

**Comment by albertocuestacanada** [comment]
First, I agree that the extra complexity is not worth the gas savings for a contract like this. The pattern would need to be way more mature. This would be interesting for a sequel to your flash-mint article, though.

Other than that, I don't see how the access controls are an issue. 

Anyone can call any `executeOnFlashMint` function anywhere, making any assumptions on the profile of the callers is not something you should do anyway. 

A contract wishing to have access control on `executeOnFlashMint` can just require that `data` includes whatever parameters needed, the interface doesn't change.

If not using access control, anyone implementing `executeOnFlashMint`, and anyone in general, must code in a way that unexpected token donations don't cause an issue. That's something most AMM developers should be doing already, anyway.

**Comment by albertocuestacanada** [comment]
@moodysalem, are you ok dropping this issue?

Flash minting is young as a pattern, and I think it is convenient to be prudent while a standard emerges.
